engines/cryomni3d/versailles/dialogs_manager.cpp

```diff
@@ -0,0 +1,367 @@
+		_engine->setupPalette(palette, 0, 256);
+	}
+
+	FontManager &fontManager = _engine->_fontManager;
+	_lastImage.create(firstFrame->w, firstFrame->h, firstFrame->format);
+	_lastImage.blitFrom(*firstFrame);
+
+	fontManager.setCurrentFont(7);
+	fontManager.setTransparentBackground(true);
+	fontManager.setForeColor(241);
+	fontManager.setLineHeight(22);
+	fontManager.setSpaceWidth(2);
+	fontManager.setCharSpacing(1);
+
+	if (_engine->showSubtitles()) {
+		Common::Rect block = settings.textRect;
+
+		unsigned int lines = fontManager.getLinesCount(text, block.width() - 8);
+		if (lines == 0) {
+			lines = 5;
+		}
+		unsigned int blockHeight = fontManager.lineHeight() * lines + 6;
+		block.setHeight(blockHeight);
+
+		if (block.bottom >= 480) {
+			block.bottom = 470;
+			warning("Dialog text is really too long");
+		}
+
+		// Make only the block area translucent inplace
+		Graphics::Surface blockSurface = _lastImage.getSubArea(block);
+		_engine->makeTranslucent(blockSurface, blockSurface);
+
+		fontManager.setSurface(&_lastImage);
+		block.grow(-4);
+		fontManager.setupBlock(block);
+		fontManager.displayBlockText(text);
+	}
+
+	g_system->copyRectToScreen(_lastImage.getPixels(), _lastImage.pitch, 0, 0, width, height);
+	g_system->updateScreen();
+
+	const Common::Rect &drawRect = settings.drawRect;
+
+	if (audioDecoder->getLength() == 0) {
+		// Empty wave file
+		delete audioDecoder;
+
+		unsigned int duration = 100 * text.size();
+		if (duration < 1000) {
+			duration = 1000;
+		}
+
+		bool skipWait = false;
+		unsigned int end = g_system->getMillis() + duration;
+		while (!g_engine->shouldQuit() && g_system->getMillis() < end && !skipWait) {
+			if (_engine->pollEvents() && _engine->checkKeysPressed(1, Common::KEYCODE_SPACE)) {
+				skipWait = true;
+			}
+		}
+	} else {
+		// Let start the show!
+		videoDecoder->start();
+
+		Audio::SoundHandle audioHandle;
+		_engine->_mixer->playStream(Audio::Mixer::kSpeechSoundType, &audioHandle, audioDecoder);
+		// We lost ownership of the audioDecoder just set it to nullptr and don't use it
+		audioDecoder = nullptr;
+
+		bool skipVideo = false;
+		while (!g_engine->shouldQuit() && _engine->_mixer->isSoundHandleActive(audioHandle) && !skipVideo) {
+			if (_engine->pollEvents() && _engine->checkKeysPressed(1, Common::KEYCODE_SPACE)) {
+				skipVideo = true;
+			}
+
+			if (videoDecoder->needsUpdate()) {
+				const Graphics::Surface *frame = videoDecoder->decodeNextFrame();
+
+				if (frame) {
+					if (videoDecoder->hasDirtyPalette()) {
+						const byte *palette = videoDecoder->getPalette();
+						_engine->setupPalette(palette, 0, 256);
+					}
+
+					// Only refresh the moving part of the animation
+					const Graphics::Surface subFrame = frame->getSubArea(drawRect);
+					g_system->copyRectToScreen(subFrame.getPixels(), subFrame.pitch, drawRect.left, drawRect.top,
+					                           subFrame.w, subFrame.h);
+				}
+			}
+			g_system->updateScreen();
```

Comment:
Another place where `delayMillis(10)` should be inserted here instead of `pollEvents()`. Otherwise, your input reaction will feel laggy.

Reply:
It might be useful to be aware that `OSystem::updateScreen` implicitly waits for the next VSync on systems with VSync enabled (most desktop platforms these days). Assuming a 60Hz display the wait can be up to 16.6 ms if a frame boundary was just passed. Maybe you want to account for that, maybe not.

Reply:
Where do you see that in code?

Reply:
If it's implicit, it's not in our code :)
This page explains nicely how wait for VSync works: https://www.khronos.org/opengl/wiki/Swap_Interval

Reply:
Oh. If OpenGL backend uses it, then it could have a negative side effects. We recommend to call updateScreen() as often as an engine wants, and then backend should just update the physical screen only when it needs to. Basically, some engines could call it as much as 200 times per second (`delayMillis(5)`).

Reply:
> We recommend to call updateScreen() as often as an engine wants, and then backend should just update the physical screen only when it needs to. Basically, some engines could call it as much as 200 times per second (delayMillis(5)).

This is not good. `updateScreen` needs to be called exactly once per engine frame. Even if nothing was drawn (to allow the backend to display its own information, to resync with vsync, possibly to allow programs that hook in, like Steam overlay, to update). Calling it more often will cause unexpected delays if wait for vsync is enabled.

Ideally the sleep time should be variable to account for the time where the CPU was stalled waiting for vsync.

 A simple game loop might be:
```cpp
const uint32 targetFrameDuration = 1000 / 60;

while (!shouldQuit()) {
    uint32 frameStart = _system->getMillis();

    doFrame();
    _system->updateScreen();

    uint32 loopElapsed = _system->getMillis() - loopStart;
    if (loopElapsed < targetFrameDuration)
        _system->delayMillis(targetFrameDuration - loopElapsed);
}
```

Reply:
What you say is exactly the opposite of the original design of the method.

Engines should not know about hardware specifics. Moreover, as you explained yourself, only a few backends have a notion of vsync, or otherwise enforced frame rate.

Thus, the code loop you put there is, in fact, belongs to the internals of updateScreen(). In there, if your backend cares about the framerate, you need to skip unneeded calls if you're being called too often.

Also, most of the engines supported by ScummVM did not care about the framerate, and would require serious refactoring, especially when the main loop of the engine is in scripts. And I understand that you're coming from the 3d engines of ResidualVM where it was modernised, thus, all of this may sound lame to you, but this is the specifics of the 80s and 90s :).